File: src/registry/binary.ts
/**
 * Varint, checksum, and byte-view primitives for LevelDB's on-disk formats.
 *
 * Deliberately knows nothing about LevelDB: both the SSTable reader and the
 * write-ahead-log reader need these, and keeping them separate is what lets
 * each of those be tested against its own format alone.
 */

export interface VarintResult {
  value: number;
  /** Offset just past the varint, so callers can thread position through. */
  next: number;
}

/** A DataView over exactly `buf`'s bytes, honouring its offset into a shared ArrayBuffer. */
export function view(buf: Uint8Array): DataView {
  return new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
}

/**
 * Reads a base-128 varint.
 *
 * Throws on a truncated buffer rather than returning a partial value: every
 * caller is parsing a file another process is actively writing, and a silently
 * short read there is the difference between "retry" and "wrong credentials".
 */
export function readVarint32(buf: Uint8Array, pos: number): VarintResult {
  let result = 0;
  let shift = 0;
  let cursor = pos;
  for (;;) {
    const byte = buf[cursor++];
    if (byte === undefined) throw new Error("varint32 ran past end of buffer");
    result |= (byte & 0x7f) << shift;
    if ((byte & 0x80) === 0) break;
    shift += 7;
    if (shift > 28) throw new Error("varint32 is longer than 5 bytes");
  }
  return { value: result >>> 0, next: cursor };
}

/**
 * Reads a 64-bit varint as a JS number.
 *
 * Block offsets and sizes are the only 64-bit varints we read, and they are
 * bounded by file size, so `number` is exact well past any real file. Values
 * beyond 2^53 are rejected rather than silently rounded.
 */
export function readVarint64(buf: Uint8Array, pos: number): VarintResult {
  let result = 0n;
  let shift = 0n;
  let cursor = pos;
  for (;;) {
    const byte = buf[cursor++];
    if (byte === undefined) throw new Error("varint64 ran past end of buffer");
    result |= BigInt(byte & 0x7f) << shift;
    if ((byte & 0x80) === 0) break;
    shift += 7n;
    if (shift > 63n) throw new Error("varint64 is longer than 10 bytes");
  }
  if (result > BigInt(Number.MAX_SAFE_INTEGER)) {
    throw new Error("varint64 exceeds the safe integer range");
  }
  return { value: Number(result), next: cursor };
}

// CRC32C (Castagnoli), reversed polynomial. LevelDB uses this, not CRC32.
const CRC32C_TABLE = (() => {
  const table = new Uint32Array(256);
  for (let n = 0; n < 256; n++) {
    let c = n;
    for (let k = 0; k < 8; k++) c = c & 1 ? 0x82f63b78 ^ (c >>> 1) : c >>> 1;
    table[n] = c >>> 0;
  }
  return table;
})();

export function crc32c(buf: Uint8Array): number {
  let crc = 0xffffffff;
  for (let i = 0; i < buf.length; i++) {
    crc = CRC32C_TABLE[(crc ^ buf[i]!) & 0xff]! ^ (crc >>> 8);
  }
  return (crc ^ 0xffffffff) >>> 0;
}

const MASK_DELTA = 0xa282ead8;

/**
 * Applies LevelDB's CRC mask: stored checksums are rotated and offset so that
 * a checksum never appears verbatim in the data it covers. The readers only
 * ever unmask; this is exported so the tests that build fixtures by hand
 * share one definition with `unmaskCrc` instead of each carrying an inverse.
 */
export function maskCrc(crc: number): number {
  const rot = ((crc >>> 15) | (crc << 17)) >>> 0;
  return (rot + MASK_DELTA) >>> 0;
}

/** Reverses `maskCrc`. */
export function unmaskCrc(masked: number): number {
  const rot = (masked - MASK_DELTA) >>> 0;
  return ((rot >>> 17) | (rot << 15)) >>> 0;
}

File: PaseoIconPackage/Sources/PaseoIconCore/Registry/Binary.swift
import Foundation

public enum BinaryError: Error, Equatable {
    /// The buffer ended inside a varint. Thrown rather than returning a partial
    /// value: every caller is parsing a file another process is actively
    /// writing, and a silently short read there is the difference between
    /// "retry" and "wrong credentials".
    case varintPastEnd(String)
    case varintTooLong(String)
}

/// Varint, checksum, and little-endian primitives for LevelDB's on-disk
/// formats. Knows nothing about LevelDB itself: both the SSTable reader and
/// the write-ahead-log reader need these, and keeping them separate is what
/// lets each of those be tested against its own format alone.
public enum Binary {
    public static func readVarint32(_ buf: [UInt8], at pos: Int) throws -> (value: UInt32, next: Int) {
        var result: UInt32 = 0
        var shift: UInt32 = 0
        var cursor = pos
        while true {
            guard cursor >= 0, cursor < buf.count else {
                throw BinaryError.varintPastEnd("varint32 ran past end of buffer")
            }
            let byte = buf[cursor]
            cursor += 1
            result |= UInt32(byte & 0x7f) << shift
            if byte & 0x80 == 0 { break }
            shift += 7
            if shift > 28 { throw BinaryError.varintTooLong("varint32 is longer than 5 bytes") }
        }
        return (result, cursor)
    }

    /// Block offsets and sizes are the only 64-bit varints read, and they are
    /// bounded by file size.
    public static func readVarint64(_ buf: [UInt8], at pos: Int) throws -> (value: UInt64, next: Int) {
        var result: UInt64 = 0
        var shift: UInt64 = 0
        var cursor = pos
        while true {
            guard cursor >= 0, cursor < buf.count else {
                throw BinaryError.varintPastEnd("varint64 ran past end of buffer")
            }
            let byte = buf[cursor]
            cursor += 1
            result |= UInt64(byte & 0x7f) << shift
            if byte & 0x80 == 0 { break }
            shift += 7
            if shift > 63 { throw BinaryError.varintTooLong("varint64 is longer than 10 bytes") }
        }
        return (result, cursor)
    }

    public static func readUInt16LE(_ buf: [UInt8], at pos: Int) -> UInt16 {
        UInt16(buf[pos]) | (UInt16(buf[pos + 1]) << 8)
    }

    public static func readUInt32LE(_ buf: [UInt8], at pos: Int) -> UInt32 {
        UInt32(buf[pos]) | (UInt32(buf[pos + 1]) << 8) | (UInt32(buf[pos + 2]) << 16) | (UInt32(buf[pos + 3]) << 24)
    }

    // CRC32C (Castagnoli), reversed polynomial. LevelDB uses this, not CRC32.
    private static let crc32cTable: [UInt32] = (0..<256).map { n -> UInt32 in
        var c = UInt32(n)
        for _ in 0..<8 { c = (c & 1) != 0 ? 0x82f6_3b78 ^ (c >> 1) : c >> 1 }
        return c
    }

    public static func crc32c(_ bytes: ArraySlice<UInt8>) -> UInt32 {
        var crc: UInt32 = 0xffff_ffff
        for byte in bytes {
            crc = crc32cTable[Int((crc ^ UInt32(byte)) & 0xff)] ^ (crc >> 8)
        }
        return crc ^ 0xffff_ffff
    }

    public static func crc32c(_ bytes: [UInt8]) -> UInt32 { crc32c(bytes[...]) }

    private static let maskDelta: UInt32 = 0xa282_ead8

    /// LevelDB's CRC mask: stored checksums are rotated and offset so that a
    /// checksum never appears verbatim in the data it covers.
    public static func maskCrc(_ crc: UInt32) -> UInt32 {
        let rot = (crc >> 15) | (crc << 17)
        return rot &+ maskDelta
    }

    public static func unmaskCrc(_ masked: UInt32) -> UInt32 {
        let rot = masked &- maskDelta
        return (rot >> 17) | (rot << 15)
    }

    /// Unsigned bytewise comparison, a shorter prefix ordering first: the
    /// order LevelDB's default comparator and Node's `Buffer.compare` share.
    public static func compare(_ a: ArraySlice<UInt8>, _ b: ArraySlice<UInt8>) -> Int {
        var ia = a.startIndex
        var ib = b.startIndex
        while ia < a.endIndex, ib < b.endIndex {
            if a[ia] != b[ib] { return a[ia] < b[ib] ? -1 : 1 }
            ia += 1
            ib += 1
        }
        if a.count == b.count { return 0 }
        return a.count < b.count ? -1 : 1
    }
}
